cpp/src/arrow/array/array_test.cc

```diff
@@ -1332,12 +1366,16 @@ TYPED_TEST(TestPrimitiveBuilder, TestAppendValuesLazyIter) {
 }
 
 TYPED_TEST(TestPrimitiveBuilder, TestAppendValuesIterConverted) {
-  DECL_T();
+  typedef typename TestFixture::CType T;
   // find type we can safely convert the tested values to and from
-  using conversion_type =
-      typename std::conditional<std::is_floating_point<T>::value, double,
-                                typename std::conditional<std::is_unsigned<T>::value,
-                                                          uint64_t, int64_t>::type>::type;
+  using conversion_type = typename std::conditional<
+      std::is_floating_point<T>::value, double,
+      typename std::conditional<
+          std::is_same<T, DayTimeIntervalType::DayMilliseconds>::value ||
+              std::is_same<T, MonthDayNanoIntervalType::MonthDayNanos>::value,
+          T,
+          typename std::conditional<std::is_unsigned<T>::value, uint64_t,
+                                    int64_t>::type>::type>::type;
```

Comment:
Hmm... can you put something on the various P* classes instead?

Reply:
done.